content/memory/memory.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package memory
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/internal/cas"
+	"oras.land/oras-go/v2/internal/descriptor"
+	"oras.land/oras-go/v2/internal/resolver"
+)
+
+// Store represents a memory based store, which implements `oras.Target`.
+type Store struct {
+	storage  content.Storage
+	resolver content.TagResolver
+	upEdges  sync.Map // map[descriptor.Descriptor]map[descriptor.Descriptor]ocispec.Descriptor
+}
+
+// New creates a new memory based store.
+func New() *Store {
+	return &Store{
+		storage:  cas.NewMemory(),
+		resolver: resolver.NewMemory(),
+	}
+}
+
+// Fetch fetches the content identified by the descriptor.
+func (s *Store) Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error) {
+	return s.storage.Fetch(ctx, target)
+}
+
+// Push pushes the content, matching the expected descriptor.
+func (s *Store) Push(ctx context.Context, expected ocispec.Descriptor, reader io.Reader) error {
+	if err := s.storage.Push(ctx, expected, reader); err != nil {
+		return err
+	}
+
+	// index up edges.
+	// there is no data consistency issue as long as deletion is not implemented
+	// for the memory store.
+	upEdgeKey := descriptor.FromOCI(expected)
+	downEdges, err := content.DownEdges(ctx, s.storage, expected)
+	if err != nil {
+		return err
+	}
+	for _, downEdge := range downEdges {
+		downEdgeKey := descriptor.FromOCI(downEdge)
+		upEdgesValue, _ := s.upEdges.LoadOrStore(downEdgeKey, &sync.Map{})
+		upEdges := upEdgesValue.(*sync.Map)
+		upEdges.Store(upEdgeKey, expected)
```

Comment:
The `upEdgeValue` variable naming is confusing. It looks like a value of the upEdges and is paired with `upEdgeKey`, but it is not...

Reply:
Let's reduce the confusion by just naming it `value`.